crates/ty_python_semantic/src/types/function.rs

```diff
@@ -0,0 +1,983 @@
+        self.has_known_decorator(db, FunctionDecorators::OVERLOAD)
+    }
+
+    fn node(self, db: &'db dyn Db, file: File) -> &'db ast::StmtFunctionDef {
+        debug_assert_eq!(
+            file,
+            self.file(db),
+            "OverloadLiteral::node() must be called with the same file as the one where \
+            the function is defined."
+        );
+
+        self.body_scope(db).node(db).expect_function()
+    }
+
+    /// Returns the [`FileRange`] of the function's name.
+    pub(crate) fn focus_range(self, db: &dyn Db) -> FileRange {
+        FileRange::new(
+            self.file(db),
+            self.body_scope(db).node(db).expect_function().name.range,
+        )
+    }
+
+    /// Returns the [`Definition`] of this function.
+    ///
+    /// ## Warning
+    ///
+    /// This uses the semantic index to find the definition of the function. This means that if the
+    /// calling query is not in the same file as this function is defined in, then this will create
+    /// a cross-module dependency directly on the full AST which will lead to cache
+    /// over-invalidation.
+    fn definition(self, db: &'db dyn Db) -> Definition<'db> {
+        let body_scope = self.body_scope(db);
+        let index = semantic_index(db, body_scope.file(db));
+        index.expect_single_definition(body_scope.node(db).expect_function())
+    }
+
+    /// Returns the overload immediately before this one in the AST. Returns `None` if there is no
+    /// previous overload.
+    fn previous_overload(self, db: &'db dyn Db) -> Option<FunctionLiteral<'db>> {
+        // The semantic model records a use for each function on the name node. This is used
+        // here to get the previous function definition with the same name.
+        let scope = self.definition(db).scope(db);
+        let use_def = semantic_index(db, scope.file(db)).use_def_map(scope.file_scope_id(db));
+        let use_id = self
+            .body_scope(db)
+            .node(db)
+            .expect_function()
+            .name
+            .scoped_use_id(db, scope);
+
+        let Symbol::Type(Type::FunctionLiteral(previous_type), Boundness::Bound) =
+            symbol_from_bindings(db, use_def.bindings_at_use(use_id))
+        else {
+            return None;
+        };
+
+        let previous_literal = previous_type.literal(db);
+        let previous_overload = previous_literal.last_definition(db);
+        if !previous_overload.is_overload(db) {
+            return None;
+        }
+
+        Some(previous_literal)
+    }
+
+    /// Typed internally-visible signature for this function.
+    ///
+    /// This represents the annotations on the function itself, unmodified by decorators and
+    /// overloads.
+    ///
+    /// These are the parameter and return types that should be used for type checking the body of
+    /// the function.
+    ///
+    /// Don't call this when checking any other file; only when type-checking the function body
+    /// scope.
```

Comment:
I think this doc comment has kind of migrated along with various refactors since I think I originally wrote it -- but looking at the current state of the code, I am not convinced it's fully accurate -- I don't think we actually use this at all for type checking the body of the function.

Reply:
That's a good point, the signature is only used when analyzing call sites. (We do use the parameter and return type annotations when type-checking the body, but we don't get those from the function's `Signature`)

Removed